### ISSUE TITLE
[AMDGPU][MC][GFX10+] Warn when disassembling a VOP3 v_cmpx* opcode with dst!=EXEC

### ISSUE THREAD

**Issue by dpreobra** [llvm/llvm-project#56727]
This is a follow up on https://reviews.llvm.org/D130345.

Joe Nash wrote:

> it also occurs to me we could use a warning when disassembling something with non-exec (not 0x7e) in the destination. The point of this was to avoid that potential error, after all, so it would help fixing suspect shaders.

**Comment by gandhi56** [comment]
@dpreobra  Should the warning be implemented in the AMDGPUDisassembler?

**Comment by dpreobra** [comment]
> @dpreobra Should the warning be implemented in the AMDGPUDisassembler?

Yes. Also, I think we should not warn when the value is 0. The message should be like "unexpected dst value".

**Comment by jayfoad** [comment]
Is there any precedent for the disassembler emitting warnings?

**Comment by dpreobra** [comment]
https://github.com/llvm/llvm-project/blob/07bfbce98872398ab58a4931ee88ccd2749a1ada/llvm/lib/Target/AMDGPU/Disassembler/AMDGPUDisassembler.cpp#L1111

**Comment by gandhi56** [comment]
https://github.com/gandhi56/llvm-project/commit/6072366fcfede0eeb4e36a8e2c46d375b01f49f4

*CommentStream does not seem to help with generating the warning. Is there a different diagnostic API I should use instead?

**Comment by dpreobra** [comment]
You are right, `CommentStream` is hardcoded to output to `nulls()` and we cannot change this. Sorry for bad advice.
But we can use `outs()` or `errs()` as a replacement, for example:

    errs() << "; Warning: " << ... ;

There is no guarantee that our warnings will be in sync with disassembler output, so the warnings should probably include the corresponding dumps.

**Comment by gandhi56** [comment]
https://reviews.llvm.org/D133334

**Comment by jayfoad** [comment]
So there isn't a precedent. It seems like target disassemblers do not emit warnings and don't have the diagnostic machinery to do it properly.

**Comment by gandhi56** [comment]
https://reviews.llvm.org/D147424

**Comment by gandhi56** [comment]
That's right. I am looking for a way to give AMDGPUDisassembler access to the stream where the instructions are emitted through annotations hopefully.

**Comment by gandhi56** [comment]
Posted an RFC: https://discourse.llvm.org/t/rfc-diagnostics-support-in-disassembler/69761

**Comment by jwanggit86** [comment]
> You are right, `CommentStream` is hardcoded to output to `nulls()` and we cannot change this. Sorry for bad advice. But we can use `outs()` or `errs()` as a replacement, for example:
> 
> ```
> errs() << "; Warning: " << ... ;
> ```
> 
> There is no guarantee that our warnings will be in sync with disassembler output, so the warnings should probably include the corresponding dumps.

After some investigation, I think issuing a warning via `errs()` in `AMDGPUDisassembler::getInstruction()` might be the best option. To include the original binary encoding in the warning msg for better understanding is also doable. Two alternative solutions I can think of are less desirable: (1) Changing the class interface such that a diagnostic object is returned requires modifying the base class. Besides, it's not easy to decide how the diagnostic object should look like? (2) Adding a new value to the return value enum, say, SoftSuccess (emulating the existing SoftFail), so that the caller can issue a warning is also possible. But the warning msg won't be precise.

**Comment by jwanggit86** [comment]
> > After some investigation, I think issuing a warning via `errs()` in `AMDGPUDisassembler::getInstruction()` might be the best
> > option.
> 
> No, this is never an acceptable end user error reporting solution.

Could you please elaborate on this and suggest an alternative?
